okhttp/src/main/java/okhttp3/internal/http/HttpEngine.java

```diff
@@ -499,6 +500,13 @@ public static boolean hasBody(Response response) {
    * content types the application is interested in.
    */
   private Request networkRequest(Request request) throws IOException {
+    if (!request.isHttps()) {
+      Authenticator authenticator = client.proxyAuthenticator();
```

Comment:
@swankjesse that's why I needed a second authenticate method, one with request as parameter. A fake response would not be a nice implementation. Maybe a mixed solution with preemptive setting in client, and adding the authenticate method with request parameter to the interface. It does not necessarily be Authenticator, maybe we can extend Authenticator by adding ProxyAuthenticator having this method.